File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;
        unordered_map<char, char> map;
        map[')'] = '(';
        map['}'] = '{';
        map[']'] = '[';
        for (auto c : s)
        {
            if (map.find(c) != map.end())
            {
                if (stk.empty() || stk.top() != map[c])
                    return false;
                stk.pop();
                continue;
            }
            
            stk.push(c);
        }
        
        return stk.empty();
    }
};

File: 0020-valid-parentheses/0020-valid-parentheses.cs
public class Solution {
    public bool IsValid(string s)
    {
        var stk = new Stack<char>();
        var open = "({[";

        foreach (var c in s)
        {
            if (open.Any(l => l == c))
            {
                stk.Push(c);
            }
            else
            {
                if (stk.Count == 0)
                    return false;

                var oc = stk.Pop();
                var expected = c switch
                {
                    ')' => '(',
                    '}' => '{',
                    ']' => '[',
                };

                if (oc != expected)
                    return false;
            }
        }

        return stk.Count == 0;
    }
}